### ISSUE TITLE
Reals and Interval

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#17360]
I started trying to implement #17280 and came up with an issue with applying the scheme to S.Reals which is currently implemented as
```python
class Reals(with_metaclass(Singleton, Interval)):
    def __new__(cls):
        return Interval.__new__(cls, -S.Infinity, S.Infinity)
```
I recall in some other places (maybe #8553 #10143 #11640) that this is considered by some contributors to be a hacky way of implementing Reals.

The problem this has with #17280 is that the args of the instance do not end up being the same as the args that are provided to the constructor.

If I was to change this what would be a better way to implement Reals? I see the motivation to subclass Interval but maybe that's not a good idea...

**Comment by jksuom** [comment]
> the args of the instance do not end up being the same as the args that are provided to the constructor.

Can you clarify? Is the problem that `left_open` and `right_open` will be True?

**Comment by oscarbenjamin** [comment]
I think that the problem is that we have the inheritance chain Basic -> ... -> Interval -> Reals and in the chain `Reals.__new__` calls `Interval.__new__` with different arguments. This means that Reals does not properly satisfy the `x == x.func(*x.args)` invariant properly - well actually it does but in a trivial way since it ignores all args after the first instance is created:
```
In [7]: S.Reals.func(1, 2, 3, 4, 5)
Out[7]: ℝ
```
I'm not sure how that happens though since `Reals.__new__` doesn't take any arguments so this should raise. Maybe that's coming from the Singleton metaclass...

I think that the line
```python
return Interval.__new__(cls, -S.Infinity, S.Infinity)
```
where cls is Reals creates an object in a way that cannot be replicated without using `__new__` since
```python
return Interval(-S.Infinity, S.Infinity)
```
gives an object of type Interval.

Perhaps it would be better if `Interval(-oo, oo)` returned Reals. Reals could still be a subclass of Interval but it would have no args and would use constant attributes in place of the start/end properties. Then it can still inherit the methods of `Interval` but can be a more well defined Basic subclass. Or maybe it would be better if the inherited Interval methods were defined in a mixin so that there isn't a direct inheritance relationship between Interval and Reals.

**Comment by oscarbenjamin** [comment]
I think this works better:
```diff
diff --git a/sympy/sets/fancysets.py b/sympy/sets/fancysets.py
index 413fe1a..015f4b7 100644
--- a/sympy/sets/fancysets.py
+++ b/sympy/sets/fancysets.py
@@ -234,8 +234,12 @@ class Reals(with_metaclass(Singleton, Interval)):
 
     ComplexRegion
     """
+    start = _inf = left = S.NegativeInfinity
+    end = _sup = right = S.Infinity
+    left_open = right_open = True
+
     def __new__(cls):
-        return Interval.__new__(cls, -S.Infinity, S.Infinity)
+        return Basic.__new__(cls)
 
     def __eq__(self, other):
         return other == Interval(-S.Infinity, S.Infinity)
```
Then Reals has no .args. It is still a subclass of Interval and can inherit the methods there but with start, end etc fixed rather than as properties that look in args.

**Comment by jksuom** [comment]
I like this construction. But does that mean that `__eq__` and `__hash__` have to be redefined?

**Comment by oscarbenjamin** [comment]
For `__eq__` the question is whether we want this behaviour:
```julia
In [1]: x = S.Reals

In [2]: y = Interval(-oo, oo)

In [3]: x
Out[3]: ℝ

In [4]: y
Out[4]: (-∞, ∞)

In [5]: x == y
Out[5]: True
```
I would say that this does not match up with the idea of == for structural equality. If we're happy with that behaviour then I think there is no need to change `__eq__/__hash__`. If we want structural equality then we can just delete both methods since they are already implemented by `Basic`.

This issue was discussed several times e.g. #7485 but then changed to the current behaviour in #9625.

**Comment by jksuom** [comment]
On possible solution is that `Interval(-oo, oo)` would be `S.Reals` in the same way as `Integer(0)` is `S.Zero`.

**Comment by oscarbenjamin** [comment]
That seems reasonable to me. We also have
```julia
In [1]: Interval(2, -1)
Out[1]: ∅
```
I note though that that idea was also rejected in some of the linked issues e.g. by @asmeurer in #7485

**Comment by jksuom** [comment]
We now have two different objects that compare equal:
```
>>> A = S.Reals
>>> B = Interval(-oo, oo)
>>> A == B
True
>>> id(A) == id(B)
False
```
I don't think that is how a singleton is expected to behave.

**Comment by oscarbenjamin** [comment]
I agree. I think that since Reals is a subclass of Interval and supports all the same methods/attributes evaluating `Interval(-oo, oo) -> Reals` is not really "evaluation" - I don't think a symbolic method could have a reason for wanting it to not happen. Under #17372 we have evaluate=False for `Interval`:
```julia
In [3]: Interval(2, -1)                                                                                                           
Out[3]: ∅

In [4]: Interval(2, -1, evaluate=False)                                                                                           
Out[4]: [2, -1]
```
so that would make it possible to use evaluate=False if wanted e.g. for printing.

**Comment by gschintgen** [comment]
While having a look at `Interval`'s `evalf` functionality, I noticed the following:
```
In [38]: S.Reals
Out[38]: ℝ

In [39]: S.Reals.n()
Out[39]: (-∞, ∞)
```
While technically not wrong, I don't see much utility in having both `Interval(-oo, oo)` and `S.Reals` for designating the very same mathematical object. So for what it's worth: I like the idea of `Interval(-oo, oo)` simply evaluating to `S.Reals`.

**Comment by asmeurer** [comment]
The way Reals and Complexes are implemented as a singletonized subclass of a more general class has always seemed odd to me. And it tends to come up in things like import time profiling (at least for Complexes) because of the way they have to instantiated as singletons. IMO both should just be distinct classes that don't subclass Interval/ComplexRegion. I also don't think they should be singletons, but it's too late to change that. 

> While technically not wrong, I don't see much utility in having both Interval(-oo, oo) and S.Reals for designating the very same mathematical object.

As long as Reals is an Interval subclass it really should work that way. Otherwise we get the odd behavior of https://github.com/sympy/sympy/issues/11640 (also mentioned here by @jksuom). If Reals were a distinct subclass from Interval, however, I don't see any reason to disallow it. The sets submodule fundamentally need to be able to represent the same set in different ways, so if that doesn't work for (-oo, oo) and R then there's a more serious problem in the module.
